File: Lab08/CO523_Lab08_E20266/concept_implementation/python/inheritance.py
# CO523 - Programming Languages | Lab 08: Object-Oriented Programming Paradigm
# Section 3.2 - Inheritance
#
# Inheritance enables a class to reuse properties and behaviours of another class.
# Manager, Developer, and Intern all inherit shared attributes and methods from Employee,
# while each adds its own specialized behavior.

class Employee:
    def __init__(self):
        # Shared attributes inherited by all subclasses
        self.name = ""
        self.id = 0
        self.salary = 0.0

    # Shared method — available to all subclasses
    def display_info(self):
        print(f"Name: {self.name}, Salary: ${self.salary}")

    def work(self):
        print(f"{self.name} is working...")


# Derived class: Manager inherits Employee and adds a bonus calculation method
class Manager(Employee):
    def calculate_bonus(self):
        return self.salary * 0.20   # Manager gets a 20% bonus


# Derived class: Developer inherits Employee and adds its own bonus calculation
class Developer(Employee):
    def calculate_bonus(self):
        return self.salary * 0.10   # Developer gets a 10% bonus


# Derived class: Intern inherits Employee and adds a learn method
class Intern(Employee):
    def learn(self):
        print(f"{self.name} is learning new skills.")


def main():
    # Manager uses inherited display_info/work and its own calculate_bonus
    mgr = Manager()
    mgr.name = "Kevin"
    mgr.id = 1023
    mgr.salary = 80000
    mgr.display_info()
    mgr.work()
    print(f"Manager Bonus: ${mgr.calculate_bonus()}")

    print()

    dev = Developer()
    dev.name = "Bob"
    dev.id = 1012
    dev.salary = 60000
    dev.display_info()
    dev.work()
    print(f"Developer Bonus: ${dev.calculate_bonus()}")

    print()

    intern = Intern()
    intern.name = "Charlie"
    intern.id = 203
    intern.salary = 20000
    intern.display_info()
    intern.work()
    intern.learn()   # Intern-specific method


if __name__ == "__main__":
    main()

File: Lab08/CO523_Lab08_E20449/concept_implementation/C_pp/inheritance.cpp
// CO523 - Programming Languages | Lab 08: Object-Oriented Programming Paradigm
// Section 3.2 - Inheritance
//
// Inheritance enables a class to reuse properties and behaviours of another class.
// It establishes parent (base) and child (derived) class relationships.
// Common logic is defined once and shared across subclasses, reducing redundancy.

#include <iostream>
#include <string>

// Base (parent) class — contains shared attributes and methods
class Employee {
public:
    std::string name;
    int id;
    double salary;

    // Shared method inherited by all derived classes
    void displayInfo() const {
        std::cout << "Name: " << name << ", Salary: $" << salary << '\n';
    }

    // Virtual method — can be overridden by subclasses
    virtual void work() const {
        std::cout << name << " is working..." << '\n';
    }

    virtual ~Employee() = default;
};

// Derived class: Manager inherits from Employee and adds a bonus method
class Manager : public Employee {
public:
    double calculateBonus() const {
        return salary * 0.20;   // Manager gets a 20% bonus
    }
};

// Derived class: Developer inherits from Employee and adds its own bonus method
class Developer : public Employee {
public:
    double calculateBonus() const {
        return salary * 0.10;   // Developer gets a 10% bonus
    }
};

// Derived class: Intern inherits from Employee and adds a learn method
class Intern : public Employee {
public:
    void learn() const {
        std::cout << name << " is learning new skills." << '\n';
    }
};

int main() {
    // Manager object uses inherited displayInfo/work and its own calculateBonus
    Manager mgr;
    mgr.name = "Kevin";
    mgr.id = 1023;
    mgr.salary = 80000;
    mgr.displayInfo();
    mgr.work();
    std::cout << "Manager Bonus: $" << mgr.calculateBonus() << '\n';

    std::cout << '\n';

    Developer dev;
    dev.name = "Bob";
    dev.id = 1012;
    dev.salary = 60000;
    dev.displayInfo();
    dev.work();
    std::cout << "Developer Bonus: $" << dev.calculateBonus() << '\n';

    std::cout << '\n';

    Intern intern;
    intern.name = "Charlie";
    intern.id = 203;
    intern.salary = 20000;
    intern.displayInfo();
    intern.work();
    intern.learn();   // Intern-specific method

    return 0;
}
